File: python/tests/test_callback_parser.py
# This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
# Copyright (c) 2024 Ian Thomas

import unittest
import sys
import os

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../src")))

from fountain_tools.callback_parser import CallbackParser

class TestParser(unittest.TestCase):

    def setUp(self):
        self.maxDiff = None  # Allow full diff output for every test case

    def _load_file(self, file_name):
        try:
            with open(f"../tests/{file_name}", "r", encoding="utf-8") as file:
                return file.read()
        except Exception as e:
            self.fail(f"Error loading {file_name}: {e}")

    def _write_title_page(self, entries):
        page = "TITLEPAGE:"
        for entry in entries:
            page += f" {entry.key}:{entry.value}"
        return page

    def test_callback_parser(self):

        out = []

        match = self._load_file("SimpleCallbackParser.txt")

        fp = CallbackParser()

        fp.onDialogue = lambda character, extension, parenthetical, line, is_dual_dialogue: out.append(f"DIALOGUE:"
            f" character:{character}"
            f" extension:{extension}"
            f" parenthetical:{parenthetical}"
            f" line:{line}"
            f" dual:{is_dual_dialogue}")

        fp.onAction = lambda text: out.append(f"ACTION: text:{text}")

        fp.onSceneHeading = lambda text, scene_number: out.append(f"HEADING: text:{text} sceneNum:{scene_number}")
          
        fp.onLyrics = lambda text: out.append(f"LYRICS: text:{text}")
        
        fp.onTransition = lambda text: out.append(f"TRANSITION: text:{text}")
        
        fp.onSection = lambda text, level: out.append(f"SECTION: level:{level} text:{text}")

        fp.onSynopsis = lambda text: out.append(f"SYNOPSIS: text:{text}")

        fp.onPageBreak = lambda: out.append("PAGEBREAK")

        fp.onTitlePage = lambda entries: out.append(self._write_title_page(entries))

        fp.ignoreBlanks = True

        fp.add_text(self._load_file('TitlePage.fountain'))
        fp.add_text(self._load_file('Sections.fountain'))
        fp.add_text(self._load_file('Character.fountain'))
        fp.add_text(self._load_file('Dialogue.fountain'))

        output = "\n".join(out)
        
        output = output.replace("None", "null")
        output = output.replace("False", "false")
        output = output.replace("True", "true")
        #print(output)
        self.assertMultiLineEqual(match, output)

File: cpp/test/test_callback_parser.cpp
#include "fountain_tools/callback_parser.h"
#include <catch_amalgamated.hpp>
#include "test_utils.h"
#include "fountain_tools/utils.h"
#include <iostream>

std::string asNull(const std::optional<std::string> value) {
    if (value.has_value())
        return value.value();
    return "null";
}

std::string asBool(const bool value) {
    return value?"true":"false";
}

TEST_CASE( "CallbackParser") {

    const std::string match = loadTestFile("SimpleCallbackParser.txt");

    std::ostringstream oss;

    Fountain::FountainCallbackParser fp;

    fp.onDialogue = [&oss](const std::string& character, const std::optional<std::string> extension, 
        const std::optional<std::string> parenthetical, const std::string&line, const bool isDualDialogue) {
        oss << "DIALOGUE:"
            << " character:" << character
            << " extension:" << asNull(extension)
            << " parenthetical:" << asNull(parenthetical)
            << " line:" << line 
            << " dual:" << asBool(isDualDialogue) << std::endl;
    };

    fp.onAction = [&oss](const std::string& text) {
        oss << "ACTION: text:" << text << std::endl;
    };

    fp.onSceneHeading = [&oss](const std::string& text, std::optional<std::string> sceneNum) {
        oss << "HEADING: text:" << text << " sceneNum:" << asNull(sceneNum) << std::endl;
    };

    fp.onLyrics = [&oss](const std::string& text) {
        oss << "LYRICS: text:" << text << std::endl;
    };

    fp.onTransition = [&oss](const std::string& text) {
        oss << "TRANSITION: text:" << text << std::endl;
    };

    fp.onSection = [&oss](const std::string& text, const int level) {
        oss << "SECTION: level:" << level << " text:" << text << std::endl;
    };

    fp.onSynopsis = [&oss](const std::string& text) {
        oss << "SYNOPSIS: text:" << text << std::endl;
    };

    fp.onPageBreak = [&oss]() {
        oss << "PAGEBREAK" << std::endl;
    };

    fp.onTitlePage = [&oss](const std::vector<Fountain::FountainCallbackParser::TitleEntry>& entries) {
        oss << "TITLEPAGE:";
        for (const auto& entry : entries) {
            oss << " " << entry.key << ":" << entry.value;
        }
        oss << std::endl;
    };

    fp.ignoreBlanks = true;

    fp.addText(loadTestFile("TitlePage.fountain"));
    fp.addText(loadTestFile("Sections.fountain"));
    fp.addText(loadTestFile("Character.fountain"));
    fp.addText(loadTestFile("Dialogue.fountain"));

    const std::string output = Fountain::trim(oss.str());
    REQUIRE(match == output);
}